packages/ui/src/components/editor/panels/Viewport/tools/PlayModeTool.tsx

```diff
@@ -47,46 +47,59 @@ import { HiOutlinePause, HiOutlinePlay } from 'react-icons/hi2'
+    removeComponent(viewerEntity, TargetCameraRotationComponent)
+    visualScriptQuery().forEach((entity) => dispatchAction(VisualScriptActions.stop({ entity })))
+    // stop all visual script logic
+  }
+  return !!entity
+}
+
+export const onStartPlayMode = () => {
+  const authState = getState(AuthState)
+  const avatarDetails = authState.user.avatar //.value
+
+  const avatarSpawnPose = getRandomSpawnPoint(Engine.instance.userID)
+  const currentScene = getComponent(getState(EditorState).rootEntity, UUIDComponent)
+
+  if (avatarDetails)
+    spawnLocalAvatarInWorld({
+      parentUUID: currentScene,
+      avatarSpawnPose,
+      avatarID: avatarDetails.id!,
+      name: authState.user.name //.value
+    })
+
+  // todo
+  // run all visual script logic
+  visualScriptQuery().forEach((entity) => dispatchAction(VisualScriptActions.execute({ entity })))
+  transformGizmoControlledQuery().forEach((entity) => removeComponent(entity, TransformGizmoControlledComponent))
+  //just remove all gizmo in the scene
+}
+
+const PlayModeTool: React.FC = () => {
   const { t } = useTranslation()
 
   const isEditing = useHookstate(getMutableState(EngineState).isEditing)
-  const authState = useHookstate(getMutableState(AuthState))
 
   const onTogglePlayMode = () => {
-    const entity = AvatarComponent.getSelfAvatarEntity()
-    if (entity) {
-      dispatchAction(WorldNetworkAction.destroyEntity({ entityUUID: getComponent(entity, UUIDComponent) }))
-      removeEntity(entity)
-      removeComponent(Engine.instance.cameraEntity, ComputedTransformComponent)
-      removeComponent(Engine.instance.cameraEntity, FollowCameraComponent)
-      removeComponent(Engine.instance.cameraEntity, TargetCameraRotationComponent)
-      getMutableState(EngineState).isEditing.set(true)
-      visualScriptQuery().forEach((entity) => dispatchAction(VisualScriptActions.stop({ entity })))
-      // stop all visual script logic
-    } else {
-      const avatarDetails = authState.user.avatar.value
-
-      const avatarSpawnPose = getRandomSpawnPoint(Engine.instance.userID)
-      const currentScene = getComponent(getState(EditorState).rootEntity, UUIDComponent)
-
-      if (avatarDetails)
-        spawnLocalAvatarInWorld({
-          parentUUID: currentScene,
-          avatarSpawnPose,
-          avatarID: avatarDetails.id!,
-          name: authState.user.name.value
-        })
-
-      // todo
-      getMutableState(EngineState).isEditing.set(false)
-      // run all visual script logic
-      visualScriptQuery().forEach((entity) => dispatchAction(VisualScriptActions.execute({ entity })))
-      transformGizmoControlledQuery().forEach((entity) => removeComponent(entity, TransformGizmoControlledComponent))
-      //just remove all gizmo in the scene
-    }
+    getMutableState(EngineState).isEditing.set(!isEditing) //getState(EngineState).isEditing)
   }
 
+  useImmediateEffect(() => {
+    !isEditing.value ? onStartPlayMode() : onStopPlayMode()
```

Comment:
useImmediateEffect is unnecessary here, and we should call stop in a callback, such that if for whatever reason PlayModeTool was to unmount, it would still reverse the effect of spawning the avatar

Reply:
by callback do you mean the return of a useEffect?

Reply:
Yes

Reply:
we need to get rid of the conditional still

`!isEditing.value ? onStartPlayMode() : onStopPlayMode()`

we can instead have 

```ts
useEffect(() => {
  if (isEditing.value) return
  onStartPlayMode() 
  return () => onStopPlayMode()
}, [isEditing])
```